load.go

```diff
@@ -22,6 +24,135 @@ func knownArg(key string) bool {
 
 const DefaultPatchStrip int = 1
 
+func (s *Source) processArgs(args map[string]string) error {
+	lex := shell.NewLex('\\')
+
+	sub := func(s *string) error {
+		updated, err := lex.ProcessWordWithMap(*s, args)
+		if err != nil {
+			return err
+		}
+		*s = updated
+		return nil
+	}
+
+	switch {
+	case s.DockerImage != nil:
+		for _, mnt := range s.DockerImage.Cmd.Mounts {
+			if err := mnt.Spec.processArgs(args); err != nil {
+				return err
+			}
+		}
+		if err := sub(&s.DockerImage.Ref); err != nil {
+			return err
+		}
+	case s.Git != nil:
+		fields := []*string{
+			&s.Git.URL,
+			&s.Git.Commit,
+		}
+		for _, f := range fields {
+			if err := sub(f); err != nil {
+				return err
+			}
+		}
+	case s.HTTPS != nil:
+		if err := sub(&s.HTTPS.URL); err != nil {
+			return err
+		}
+	case s.Context != nil:
+		if err := sub(&s.Context.Name); err != nil {
+			return err
+		}
+	case s.Build != nil:
+		if err := s.Build.Source.processArgs(args); err != nil {
+			return err
+		}
+
+		fields := []*string{
+			&s.Build.DockerFile,
+			&s.Build.Target,
+		}
+		for _, f := range fields {
+			if err := sub(f); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func fillDefaults(s *Source) {
+	switch {
+	case s.DockerImage != nil:
+		for _, mnt := range s.DockerImage.Cmd.Mounts {
+			fillDefaults(&mnt.Spec)
+		}
+	case s.Git != nil:
+	case s.HTTPS != nil:
+	case s.Context != nil:
+		if s.Context.Name == "" {
+			s.Context.Name = dockerui.DefaultLocalNameContext
+		}
+		if s.Path == "" {
+			s.Path = "."
+		}
+	case s.Build != nil:
+		fillDefaults(&s.Build.Source)
+	}
+}
+
+func (s *Source) validate() error {
+	count := 0
+	var errs error
+
+	if s.DockerImage != nil {
+		for _, mnt := range s.DockerImage.Cmd.Mounts {
+			if err := mnt.Spec.validate(); err != nil {
+				errs = goerrors.Join(errs, err)
+			}
+		}
+
+		count++
+	}
+	if s.Git != nil {
+		count++
+	}
+	if s.HTTPS != nil {
+		count++
+	}
+	if s.Context != nil {
+		count++
+	}
+	if s.Build != nil {
+		if s.Build.Source.Build != nil {
+			errs = goerrors.Join(errs, fmt.Errorf("build sources cannot be recursive"))
+		}
+
+		if s.Build.DockerFile != "" && s.Build.Inline != "" {
+			errs = goerrors.Join(errs, fmt.Errorf("build sources may use either `dockerfile` or `inline`, but not both"))
+		}
+
+		if err := s.Build.Source.validate(); err != nil {
+			errs = goerrors.Join(errs, err)
+		}
+
+		count++
+	}
+
+	switch count {
+	case 0:
+		errs = goerrors.Join(errs, fmt.Errorf("no non-nil source variant"))
+	case 1:
+		// success condition
+	default:
+		errs = goerrors.Join(errs, fmt.Errorf("more than one source variant defined"))
+	}
+
+	return errs
```

Comment:
Ah, the reason I didn't do this is because I check the mounts **and** the source types. If there are multiple errors I thought it might be useful to return all of them. Instead, I've changed it to report errors in the mounts immediately, and then added the above.

Reply:
(done)